edx_exams/apps/api/v1/views.py

```diff
@@ -271,3 +274,116 @@ class ProctoringProvidersView(ListAPIView):
     model = ProctoringProvider
     serializer_class = ProctoringProviderSerializer
     queryset = ProctoringProvider.objects.all()
+
+
+class ExamAccessTicketsView(APIView):
+    """
+    View to create signed exam access tickets for a specific user and exam.
+
+    Given an exam_id and user (in request), this view will either grant access
+    as an exam access ticket or not grant access. Access is granted if there is an
+    existing exam attempt or if the exam is past its due date.
+
+    HTTP GET
+    Provides an Exam Access Ticket as a cookie if access is granted.
+    **GET data Parameters**
+        * exam_id: This is the id of the exam that the user is requesting access to
+    **Exceptions**
+        * HTTP_403_FORBIDDEN
+    """
+
+    authentication_classes = (JwtAuthentication,)
+    permission_classes = (StaffUserOrReadOnlyPermissions,)
+
+    @classmethod
+    def get_current_exam_attempt(cls, user, exam):
+        """
+        Given a user and exam, get the user's latest exam attempt, if exists.
+        """
+        try:
+            exam_attempt = ExamAttempt.objects.filter(user=user, exam=exam).latest('created')
+        except ObjectDoesNotExist:
+            exam_attempt = None
+        return exam_attempt
+
+    @classmethod
+    def set_expiration_window(cls, exam_attempt, default_exp_seconds):
+        """
+        Set exam access ticket expiration window.
+
+        Use default exp window or the time to the end of the exam attempt,
+        whichever is shorter.
+        """
+        secs_remaining = get_exam_attempt_time_remaining(exam_attempt.start_time,
+                                                         exam_attempt.allowed_time_limit_mins)
+
+        # If not past due and attempt time remaining is less than default, set window to remaining
+        if 0 <= secs_remaining < default_exp_seconds:
+            exp_window = secs_remaining
+        else:
+            exp_window = default_exp_seconds
+
+        return exp_window
+
+    @classmethod
+    def get_response(cls, exam, user):
+        """
+        Get response, with exam access ticket if access granted.
+
+        403 error if access is not granted.
+        """
+        claims = {"course_id": exam.course_id, "content_id": exam.content_id}
+        expiration_window = 60
+        exam_attempt = cls.get_current_exam_attempt(user, exam)
+        cookie_name = "exam_access_ticket"
+
+        data = {}
+        grant_access = True
+        response_status = status.HTTP_200_OK
+
+        # If exam attempt exists for user, then grant exam access
```

Comment:
Having a hard time following this block. If the exam attempt doesn't exist at all maybe we could just short circuit and return a 404 first? That might make structuring the rest of this easier to read.

What about created, ready, submitted, etc attempts? Seems to me they just fall through all of these cases and get access by default.

Reply:
Ah maybe I didn't understand the logic correctly here then. I thought that if the exam attempt doesn't exist, but the exam is past due, that the person should still get access?

Oh, yeah. I didn't really realize those were potential cases. If they're created, ready, submitted, what's the expected behavior? Should we only give access if started, or are some of these other cases also ones where access should be granted?

Reply:
oh yeah thinking about this more my original comment doesn't even cover the whole scope of this. We also differentiate behavior between timed and proctored exams. See https://2u-internal.atlassian.net/wiki/spaces/PT/pages/15440919/Exam+Visibility+Behavior

You can also take a look at the code that deals with this today https://github.com/openedx/edx-proctoring/blob/0ae746c253db5910cd6f91abf60e19771102130c/edx_proctoring/api.py#L2958. There's a bunch of other complexities beyond 'does the learner have access' in that code though.

Honestly, I'm happy to make the decision to just simplify this and get rid of the proctored vs. timed differentiation. We can always add that back later if it creates problems but there's been a desire by the support team to change this anyway.

My suggestion:
- Before the due date exams should only ever return in access token if the attempt is in the 'started' state.
- After the due date exams would allow an access token if the attempt is 'submitted' (if timed) or 'verified' (if proctored) after due. This only applies if exam.hide_after_due is False, otherwise no access token

Reply:
This is a more basic question about exams, but do we need to consider the case where it is a practice exam and if so what would the access look like? Would we always grant access? Or is this case not relevant here? If not, how are we certain that we won't get practice exams for for the exam access endpoint?

The main reason I ask is because I noticed in the Exam Visibility Behavior doc that all special exams (proctored and timed) have due dates. So I am wondering if I need to actually check if the exam has a due date first or if I can just check what the due date is relative to now.

Reply:
I don't think a practice exam would change the way this works since it should behave the same as it's proctored or timed counterpart, it's just not graded. As far as due date, we should check against that because it is possible an exam may not actually have a due date. I think the doc just doesn't call it out but those exams are just perpetually 'before the due date'.